File: nodejs/test/e2e/session_fs.test.ts
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *--------------------------------------------------------------------------------------------*/

import { SessionCompactionCompleteEvent } from "@github/copilot/sdk";
import { MemoryProvider, VirtualProvider } from "@platformatic/vfs";
import { describe, expect, it, onTestFinished } from "vitest";
import { CopilotClient } from "../../src/client.js";
import { SessionFsHandler } from "../../src/generated/rpc.js";
import {
    approveAll,
    CopilotSession,
    defineTool,
    SessionEvent,
    type SessionFsConfig,
} from "../../src/index.js";
import { createSdkTestContext } from "./harness/sdkTestContext.js";

describe("Session Fs", async () => {
    // Single provider for the describe block — session IDs are unique per test,
    // so no cross-contamination between tests.
    const provider = new MemoryProvider();
    const createSessionFsHandler = (session: CopilotSession) =>
        createTestSessionFsHandler(session, provider);

    // Helpers to build session-namespaced paths for direct provider assertions
    const p = (sessionId: string, path: string) =>
        `/${sessionId}${path.startsWith("/") ? path : "/" + path}`;

    const { copilotClient: client, env } = await createSdkTestContext({
        copilotClientOptions: { sessionFs: sessionFsConfig },
    });

    it("should route file operations through the session fs provider", async () => {
        const session = await client.createSession({
            onPermissionRequest: approveAll,
            createSessionFsHandler,
        });

        const msg = await session.sendAndWait({ prompt: "What is 100 + 200?" });
        expect(msg?.data.content).toContain("300");
        await session.disconnect();

        const buf = await provider.readFile(p(session.sessionId, "/session-state/events.jsonl"));
        const content = buf.toString("utf8");
        expect(content).toContain("300");
    });

    it("should load session data from fs provider on resume", async () => {
        const session1 = await client.createSession({
            onPermissionRequest: approveAll,
            createSessionFsHandler,
        });
        const sessionId = session1.sessionId;

        const msg = await session1.sendAndWait({ prompt: "What is 50 + 50?" });
        expect(msg?.data.content).toContain("100");
        await session1.disconnect();

        // The events file should exist before resume
        expect(await provider.exists(p(sessionId, "/session-state/events.jsonl"))).toBe(true);

        const session2 = await client.resumeSession(sessionId, {
            onPermissionRequest: approveAll,
            createSessionFsHandler,
        });

        // Send another message to verify the session is functional after resume
        const msg2 = await session2.sendAndWait({ prompt: "What is that times 3?" });
        await session2.disconnect();
        expect(msg2?.data.content).toContain("300");
    });

    it("should reject setProvider when sessions already exist", async () => {
        const client = new CopilotClient({
            useStdio: false, // Use TCP so we can connect from a second client
            env,
        });
        await client.createSession({ onPermissionRequest: approveAll, createSessionFsHandler });

        // Get the port the first client's runtime is listening on
        const port = (client as unknown as { actualPort: number }).actualPort;

        // Second client tries to connect with a session fs — should fail
        // because sessions already exist on the runtime.
        const client2 = new CopilotClient({
            env,
            logLevel: "error",
            cliUrl: `localhost:${port}`,
            sessionFs: sessionFsConfig,
        });
        onTestFinished(() => client2.forceStop());

        await expect(client2.start()).rejects.toThrow();
    });

    it("should map large output handling into sessionFs", async () => {
        const suppliedFileContent = "x".repeat(100_000);
        const session = await client.createSession({
            onPermissionRequest: approveAll,
            createSessionFsHandler,
            tools: [
                defineTool("get_big_string", {
                    description: "Returns a large string",
                    handler: async () => suppliedFileContent,
                }),
            ],
        });

        await session.sendAndWait({
            prompt: "Call the get_big_string tool and reply with the word DONE only.",
        });

        // The tool result should reference a temp file under the session state path
        const messages = await session.getMessages();
        const toolResult = findToolCallResult(messages, "get_big_string");
        expect(toolResult).toContain("/session-state/temp/");
        const filename = toolResult?.match(/(\/session-state\/temp\/[^\s]+)/)?.[1];
        expect(filename).toBeDefined();

        // Verify the file was written with the correct content via the provider
        const fileContent = await provider.readFile(p(session.sessionId, filename!), "utf8");
        expect(fileContent).toBe(suppliedFileContent);
    });

    it("should write workspace metadata via sessionFs", async () => {
        const session = await client.createSession({
            onPermissionRequest: approveAll,
            createSessionFsHandler,
        });

        const msg = await session.sendAndWait({ prompt: "What is 7 * 8?" });
        expect(msg?.data.content).toContain("56");

        // WorkspaceManager should have created workspace.yaml via sessionFs
        const workspaceYamlPath = p(session.sessionId, "/session-state/workspace.yaml");
        await expect.poll(() => provider.exists(workspaceYamlPath)).toBe(true);
        const yaml = await provider.readFile(workspaceYamlPath, "utf8");
        expect(yaml).toContain("id:");

        // Checkpoint index should also exist
        const indexPath = p(session.sessionId, "/session-state/checkpoints/index.md");
        await expect.poll(() => provider.exists(indexPath)).toBe(true);

        await session.disconnect();
    });

    it("should persist plan.md via sessionFs", async () => {
        const session = await client.createSession({
            onPermissionRequest: approveAll,
            createSessionFsHandler,
        });

        // Write a plan via the session RPC
        await session.sendAndWait({ prompt: "What is 2 + 3?" });
        await session.rpc.plan.update({ content: "# Test Plan\n\nThis is a test." });

        const planPath = p(session.sessionId, "/session-state/plan.md");
        await expect.poll(() => provider.exists(planPath)).toBe(true);
        const content = await provider.readFile(planPath, "utf8");
        expect(content).toContain("# Test Plan");

        await session.disconnect();
    });

    it("should succeed with compaction while using sessionFs", async () => {
        const session = await client.createSession({
            onPermissionRequest: approveAll,
            createSessionFsHandler,
        });

        let compactionEvent: SessionCompactionCompleteEvent | undefined;
        session.on("session.compaction_complete", (evt) => (compactionEvent = evt));

        await session.sendAndWait({ prompt: "What is 2+2?" });

        const eventsPath = p(session.sessionId, "/session-state/events.jsonl");
        await expect.poll(() => provider.exists(eventsPath)).toBe(true);
        const contentBefore = await provider.readFile(eventsPath, "utf8");
        expect(contentBefore).not.toContain("checkpointNumber");

        await session.rpc.history.compact();
        await expect.poll(() => compactionEvent).toBeDefined();
        expect(compactionEvent!.data.success).toBe(true);

        // Verify the events file was rewritten with a checkpoint via sessionFs
        await expect
            .poll(() => provider.readFile(eventsPath, "utf8"))
            .toContain("checkpointNumber");
    });
});

function findToolCallResult(messages: SessionEvent[], toolName: string): string | undefined {
    for (const m of messages) {
        if (m.type === "tool.execution_complete") {
            if (findToolName(messages, m.data.toolCallId) === toolName) {
                return m.data.result?.content;
            }
        }
    }
}

function findToolName(messages: SessionEvent[], toolCallId: string): string | undefined {
    for (const m of messages) {
        if (m.type === "tool.execution_start" && m.data.toolCallId === toolCallId) {
            return m.data.toolName;
        }
    }
}

const sessionFsConfig: SessionFsConfig = {
    initialCwd: "/",
    sessionStatePath: "/session-state",
    conventions: "posix",
};

function createTestSessionFsHandler(
    session: CopilotSession,
    provider: VirtualProvider
): SessionFsHandler {
    const sp = (sessionId: string, path: string) =>
        `/${sessionId}${path.startsWith("/") ? path : "/" + path}`;

    function mapError(err: unknown): { code: "ENOENT" | "UNKNOWN"; message?: string } {
        const e = err as NodeJS.ErrnoException;
        if (e.code === "ENOENT") return { code: "ENOENT", message: e.message };
        return { code: "UNKNOWN", message: e.message ?? String(err) };
    }

    return {
        readFile: async ({ path }) => {
            try {
                const content = await provider.readFile(sp(session.sessionId, path), "utf8");
                return { content: content as string };
            } catch (err) {
                return { content: "", error: mapError(err) };
            }
        },
        writeFile: async ({ path, content }) => {
            try {
                await provider.writeFile(sp(session.sessionId, path), content);
                return {};
            } catch (err) {
                return { error: mapError(err) };
            }
        },
        appendFile: async ({ path, content }) => {
            try {
                await provider.appendFile(sp(session.sessionId, path), content);
                return {};
            } catch (err) {
                return { error: mapError(err) };
            }
        },
        exists: async ({ path }) => {
            return { exists: await provider.exists(sp(session.sessionId, path)) };
        },
        stat: async ({ path }) => {
            try {
                const st = await provider.stat(sp(session.sessionId, path));
                return {
                    isFile: st.isFile(),
                    isDirectory: st.isDirectory(),
                    size: st.size,
                    mtime: new Date(st.mtimeMs).toISOString(),
                    birthtime: new Date(st.birthtimeMs).toISOString(),
                };
            } catch (err) {
                return { isFile: false, isDirectory: false, size: 0, mtime: new Date().toISOString(), birthtime: new Date().toISOString(), error: mapError(err) };
            }
        },
        mkdir: async ({ path, recursive, mode }) => {
            try {
                await provider.mkdir(sp(session.sessionId, path), {
                    recursive: recursive ?? false,
                    mode,
                });
                return {};
            } catch (err) {
                return { error: mapError(err) };
            }
        },
        readdir: async ({ path }) => {
            try {
                const entries = await provider.readdir(sp(session.sessionId, path));
                return { entries: entries as string[] };
            } catch (err) {
                return { entries: [], error: mapError(err) };
            }
        },
        readdirWithTypes: async ({ path }) => {
            try {
                const names = (await provider.readdir(sp(session.sessionId, path))) as string[];
                const entries = await Promise.all(
                    names.map(async (name) => {
                        const st = await provider.stat(sp(session.sessionId, `${path}/${name}`));
                        return {
                            name,
                            type: st.isDirectory() ? ("directory" as const) : ("file" as const),
                        };
                    })
                );
                return { entries };
            } catch (err) {
                return { entries: [], error: mapError(err) };
            }
        },
        rm: async ({ path }) => {
            try {
                await provider.unlink(sp(session.sessionId, path));
                return {};
            } catch (err) {
                return { error: mapError(err) };
            }
        },
        rename: async ({ src, dest }) => {
            try {
                await provider.rename(sp(session.sessionId, src), sp(session.sessionId, dest));
                return {};
            } catch (err) {
                return { error: mapError(err) };
            }
        },
    };
}

File: go/internal/e2e/session_fs_test.go
package e2e

import (
	"fmt"
	"sort"
	"strings"
	"sync"
	"testing"
	"time"

	copilot "github.com/github/copilot-sdk/go"
	"github.com/github/copilot-sdk/go/internal/e2e/testharness"
)

// inMemoryFS is a simple in memory filesystem for testing.
type inMemoryFS struct {
	mu    sync.Mutex
	files map[string]string
	dirs  map[string]bool
	mtime map[string]time.Time
}

func newInMemoryFS() *inMemoryFS {
	return &inMemoryFS{
		files: make(map[string]string),
		dirs:  map[string]bool{"/": true},
		mtime: make(map[string]time.Time),
	}
}

func (fs *inMemoryFS) ensureParents(p string) {
	parts := strings.Split(p, "/")
	for i := 1; i < len(parts)-1; i++ {
		dir := strings.Join(parts[:i+1], "/")
		fs.dirs[dir] = true
	}
}

// sessionFsHandler adapts the in memory FS for a specific session.
type sessionFsHandler struct {
	sessionID string
	fs        *inMemoryFS
}

func (h *sessionFsHandler) sp(p string) string {
	if strings.HasPrefix(p, "/") {
		return "/" + h.sessionID + p
	}
	return "/" + h.sessionID + "/" + p
}

func (h *sessionFsHandler) ReadFile(params copilot.SessionFsReadFileParams) (*copilot.SessionFsReadFileResult, error) {
	h.fs.mu.Lock()
	defer h.fs.mu.Unlock()
	content, ok := h.fs.files[h.sp(params.Path)]
	if !ok {
		return nil, fmt.Errorf("file not found: %s", params.Path)
	}
	return &copilot.SessionFsReadFileResult{Content: content}, nil
}

func (h *sessionFsHandler) WriteFile(params copilot.SessionFsWriteFileParams) error {
	h.fs.mu.Lock()
	defer h.fs.mu.Unlock()
	p := h.sp(params.Path)
	h.fs.ensureParents(p)
	h.fs.files[p] = params.Content
	h.fs.mtime[p] = time.Now()
	return nil
}

func (h *sessionFsHandler) AppendFile(params copilot.SessionFsAppendFileParams) error {
	h.fs.mu.Lock()
	defer h.fs.mu.Unlock()
	p := h.sp(params.Path)
	h.fs.ensureParents(p)
	h.fs.files[p] += params.Content
	h.fs.mtime[p] = time.Now()
	return nil
}

func (h *sessionFsHandler) Exists(params copilot.SessionFsExistsParams) (*copilot.SessionFsExistsResult, error) {
	h.fs.mu.Lock()
	defer h.fs.mu.Unlock()
	p := h.sp(params.Path)
	_, fileOk := h.fs.files[p]
	_, dirOk := h.fs.dirs[p]
	return &copilot.SessionFsExistsResult{Exists: fileOk || dirOk}, nil
}

func (h *sessionFsHandler) Stat(params copilot.SessionFsStatParams) (*copilot.SessionFsStatResult, error) {
	h.fs.mu.Lock()
	defer h.fs.mu.Unlock()
	p := h.sp(params.Path)
	if content, ok := h.fs.files[p]; ok {
		mt := h.fs.mtime[p]
		return &copilot.SessionFsStatResult{
			IsFile:      true,
			IsDirectory: false,
			Size:        int64(len(content)),
			Mtime:       mt.Format(time.RFC3339Nano),
			Birthtime:   mt.Format(time.RFC3339Nano),
		}, nil
	}
	if h.fs.dirs[p] {
		return &copilot.SessionFsStatResult{
			IsFile:      false,
			IsDirectory: true,
			Mtime:       time.Now().Format(time.RFC3339Nano),
			Birthtime:   time.Now().Format(time.RFC3339Nano),
		}, nil
	}
	return nil, fmt.Errorf("path not found: %s", params.Path)
}

func (h *sessionFsHandler) Mkdir(params copilot.SessionFsMkdirParams) error {
	h.fs.mu.Lock()
	defer h.fs.mu.Unlock()
	p := h.sp(params.Path)
	if params.Recursive != nil && *params.Recursive {
		h.fs.ensureParents(p + "/x")
	}
	h.fs.dirs[p] = true
	return nil
}

func (h *sessionFsHandler) Readdir(params copilot.SessionFsReaddirParams) (*copilot.SessionFsReaddirResult, error) {
	h.fs.mu.Lock()
	defer h.fs.mu.Unlock()
	p := h.sp(params.Path)
	if !strings.HasSuffix(p, "/") {
		p += "/"
	}
	entries := map[string]bool{}
	for k := range h.fs.files {
		if strings.HasPrefix(k, p) && len(k) > len(p) {
			rest := k[len(p):]
			if idx := strings.Index(rest, "/"); idx >= 0 {
				entries[rest[:idx]] = true
			} else {
				entries[rest] = true
			}
		}
	}
	for k := range h.fs.dirs {
		if strings.HasPrefix(k, p) && len(k) > len(p) {
			rest := k[len(p):]
			if idx := strings.Index(rest, "/"); idx >= 0 {
				entries[rest[:idx]] = true
			} else {
				entries[rest] = true
			}
		}
	}
	result := make([]string, 0, len(entries))
	for e := range entries {
		result = append(result, e)
	}
	sort.Strings(result)
	return &copilot.SessionFsReaddirResult{Entries: result}, nil
}

func (h *sessionFsHandler) ReaddirWithTypes(params copilot.SessionFsReaddirWithTypesParams) (*copilot.SessionFsReaddirWithTypesResult, error) {
	dirResult, err := h.Readdir(copilot.SessionFsReaddirParams{SessionID: params.SessionID, Path: params.Path})
	if err != nil {
		return nil, err
	}
	p := h.sp(params.Path)
	if !strings.HasSuffix(p, "/") {
		p += "/"
	}
	h.fs.mu.Lock()
	defer h.fs.mu.Unlock()
	var entries []copilot.SessionFsDirEntry
	for _, name := range dirResult.Entries {
		full := p + name
		entryType := "file"
		if h.fs.dirs[full] {
			entryType = "directory"
		} else {
			// Check if any file has this as prefix (implicit directory)
			for k := range h.fs.files {
				if strings.HasPrefix(k, full+"/") {
					entryType = "directory"
					break
				}
			}
		}
		entries = append(entries, copilot.SessionFsDirEntry{Name: name, Type: entryType})
	}
	return &copilot.SessionFsReaddirWithTypesResult{Entries: entries}, nil
}

func (h *sessionFsHandler) Rm(params copilot.SessionFsRmParams) error {
	h.fs.mu.Lock()
	defer h.fs.mu.Unlock()
	p := h.sp(params.Path)
	delete(h.fs.files, p)
	delete(h.fs.dirs, p)
	delete(h.fs.mtime, p)
	return nil
}

func (h *sessionFsHandler) Rename(params copilot.SessionFsRenameParams) error {
	h.fs.mu.Lock()
	defer h.fs.mu.Unlock()
	src := h.sp(params.Src)
	dest := h.sp(params.Dest)
	if content, ok := h.fs.files[src]; ok {
		h.fs.ensureParents(dest)
		h.fs.files[dest] = content
		h.fs.mtime[dest] = h.fs.mtime[src]
		delete(h.fs.files, src)
		delete(h.fs.mtime, src)
	}
	return nil
}

func TestSessionFs(t *testing.T) {
	ctx := testharness.NewTestContext(t)

	// Shared in memory filesystem across tests
	memFS := newInMemoryFS()

	client := ctx.NewClient(func(opts *copilot.ClientOptions) {
		opts.SessionFs = &copilot.SessionFsConfig{
			InitialCwd:       "/",
			SessionStatePath: "/session-state",
			Conventions:      "posix",
		}
	})
	t.Cleanup(func() { client.ForceStop() })

	t.Run("should route file operations through the session fs provider", func(t *testing.T) {
		ctx.ConfigureForTest(t)

		session, err := client.CreateSession(t.Context(), &copilot.SessionConfig{
			OnPermissionRequest: copilot.PermissionHandler.ApproveAll,
			CreateSessionFsHandler: func(s *copilot.Session) copilot.SessionFsHandler {
				return &sessionFsHandler{sessionID: s.SessionID, fs: memFS}
			},
		})
		if err != nil {
			t.Fatalf("Failed to create session: %v", err)
		}

		msg, err := session.SendAndWait(t.Context(), copilot.MessageOptions{Prompt: "What is 100 + 200?"})
		if err != nil {
			t.Fatalf("Failed to send message: %v", err)
		}
		if msg == nil || msg.Data.Content == nil || !strings.Contains(*msg.Data.Content, "300") {
			t.Fatalf("Expected response containing '300', got: %v", msg)
		}
		session.Disconnect()

		// Verify the events file was written through our provider
		eventsPath := "/" + session.SessionID + "/session-state/events.jsonl"
		memFS.mu.Lock()
		content, ok := memFS.files[eventsPath]
		memFS.mu.Unlock()
		if !ok {
			t.Fatal("Expected events.jsonl to exist in in memory filesystem")
		}
		if !strings.Contains(content, "300") {
			t.Errorf("Expected events.jsonl to contain '300', got: %s", content[:min(200, len(content))])
		}
	})

	t.Run("should load session data from fs provider on resume", func(t *testing.T) {
		ctx.ConfigureForTest(t)

		session1, err := client.CreateSession(t.Context(), &copilot.SessionConfig{
			OnPermissionRequest: copilot.PermissionHandler.ApproveAll,
			CreateSessionFsHandler: func(s *copilot.Session) copilot.SessionFsHandler {
				return &sessionFsHandler{sessionID: s.SessionID, fs: memFS}
			},
		})
		if err != nil {
			t.Fatalf("Failed to create session: %v", err)
		}
		sessionID := session1.SessionID

		msg, err := session1.SendAndWait(t.Context(), copilot.MessageOptions{Prompt: "What is 50 + 50?"})
		if err != nil {
			t.Fatalf("Failed to send message: %v", err)
		}
		if msg == nil || msg.Data.Content == nil || !strings.Contains(*msg.Data.Content, "100") {
			t.Fatalf("Expected response containing '100', got: %v", msg)
		}
		session1.Disconnect()

		// Verify events file exists
		eventsPath := "/" + sessionID + "/session-state/events.jsonl"
		memFS.mu.Lock()
		_, exists := memFS.files[eventsPath]
		memFS.mu.Unlock()
		if !exists {
			t.Fatal("Expected events.jsonl to exist before resume")
		}

		// Resume the session
		session2, err := client.ResumeSession(t.Context(), sessionID, &copilot.ResumeSessionConfig{
			OnPermissionRequest: copilot.PermissionHandler.ApproveAll,
			CreateSessionFsHandler: func(s *copilot.Session) copilot.SessionFsHandler {
				return &sessionFsHandler{sessionID: s.SessionID, fs: memFS}
			},
		})
		if err != nil {
			t.Fatalf("Failed to resume session: %v", err)
		}

		msg2, err := session2.SendAndWait(t.Context(), copilot.MessageOptions{Prompt: "What is that times 3?"})
		if err != nil {
			t.Fatalf("Failed to send message: %v", err)
		}
		session2.Disconnect()
		if msg2 == nil || msg2.Data.Content == nil || !strings.Contains(*msg2.Data.Content, "300") {
			t.Fatalf("Expected response containing '300', got: %v", msg2)
		}
	})
}
